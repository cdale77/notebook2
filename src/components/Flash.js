import React from "react";
import PropTypes from "prop-types";

const Flash = ({ flashType, message, closeFlash }) => {
  const isFlashPresent = message => {
    return message !== "";
  };

  const setTimer = () => {
    setTimeout(closeFlash, 4000);
  };

  if (isFlashPresent(message)) {
    setTimer();
    return (
      <div id="flash" className={"flash-" + flashType}>
        <div className="flash-message">{message}</div>
        <div className="close-button">
          <button onClick={closeFlash} className="close-flash btn-small">
            X
          </button>
        </div>
      </div>
    );
  } else {
    return <div />;
  }
};

Flash.propTypes = {
  flashType: PropTypes.string,
  message: PropTypes.string,
  closeFlash: PropTypes.func.isRequired
};

Flash.defaultProps = {
  message: ""
};

export default Flash;
