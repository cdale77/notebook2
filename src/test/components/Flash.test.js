import React from "react";
import Renderer from "react-test-renderer";
import TestUtils from "react-dom/test-utils";
import TestWrapper from "../helpers/TestWrapper";
import Flash from "../../components/Flash";

describe("Flash", () => {
  describe("rendering", () => {
    it("should display when the message prop present", () => {
      const component = Renderer.create(
        <Flash closeFlash={jest.fn()} message="It worked!" />
      );

      const tree = component.toJSON();
      expect(tree).toMatchSnapshot();
    });

    it("should not display when the message is blank", () => {
      const component = Renderer.create(
        <Flash closeFlash={jest.fn()} message="" />
      );

      const tree = component.toJSON();
      expect(tree).toMatchSnapshot();
    });

    it("should not display when the message is not provided", () => {
      const component = Renderer.create(<Flash closeFlash={jest.fn()} />);

      const tree = component.toJSON();
      expect(tree).toMatchSnapshot();
    });

    it("should display when the message prop present", () => {
      const component = Renderer.create(
        <Flash closeFlash={jest.fn()} message="It worked!" />
      );

      const tree = component.toJSON();
      expect(tree).toMatchSnapshot();
    });
  });

  describe("closing", () => {
    let callBack, spy, component;

    beforeEach(() => {
      jest.useFakeTimers();

      callBack = {
        closeFlash: () => {}
      };

      spy = jest.spyOn(callBack, "closeFlash");

      component = TestUtils.renderIntoDocument(
        <TestWrapper>
          <Flash closeFlash={callBack.closeFlash} message="It worked!" />
        </TestWrapper>
      );
    });

    it("should close after 2 seconds", () => {
      expect(spy).not.toHaveBeenCalled();

      // Run out timers
      jest.runAllTimers();

      expect(spy).toHaveBeenCalled();
      expect(callBack.closeFlash.mock.calls.length).toEqual(1);
    });

    it("should close when the close icon is clicked", () => {
      const closeIcon = TestUtils.findRenderedDOMComponentWithClass(
        component,
        "close-flash"
      );

      TestUtils.Simulate.click(closeIcon);

      expect(spy).toHaveBeenCalled();
      expect(callBack.closeFlash.mock.calls.length).toEqual(1);
    });
  });
});
