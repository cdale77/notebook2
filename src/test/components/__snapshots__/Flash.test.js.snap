// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Flash rendering should display when the message prop present 1`] = `
<div
  className="flash-undefined"
  id="flash"
>
  <div
    className="flash-message"
  >
    It worked!
  </div>
  <div
    className="close-button"
  >
    <button
      className="close-flash btn-small"
      onClick={[Function]}
    >
      X
    </button>
  </div>
</div>
`;

exports[`Flash rendering should display when the message prop present 2`] = `
<div
  className="flash-undefined"
  id="flash"
>
  <div
    className="flash-message"
  >
    It worked!
  </div>
  <div
    className="close-button"
  >
    <button
      className="close-flash btn-small"
      onClick={[Function]}
    >
      X
    </button>
  </div>
</div>
`;

exports[`Flash rendering should not display when the message is blank 1`] = `<div />`;

exports[`Flash rendering should not display when the message is not provided 1`] = `<div />`;
